FirebaseRemoteConfig/SwiftNew/ConfigSettings.swift

```diff
@@ -0,0 +1,579 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import GoogleUtilities
+
+// TODO(ncooke3): Once Obj-C tests are ported, all `public` access modifers can be removed.
+
+private let kRCNGroupPrefix = "frc.group."
+private let kRCNUserDefaultsKeyNamelastETag = "lastETag"
+private let kRCNUserDefaultsKeyNameLastSuccessfulFetchTime = "lastSuccessfulFetchTime"
+private let kRCNAnalyticsFirstOpenTimePropertyName = "_fot"
+private let kRCNExponentialBackoffMinimumInterval = 60 * 2 // 2 mins.
+private let kRCNExponentialBackoffMaximumInterval = 60 * 60 * 4 // 4 hours.
+
+let RCNHTTPDefaultConnectionTimeout: TimeInterval = 60
+
+/// This internal class contains a set of variables that are unique among all the config instances.
+/// It also handles all metadata. This class is not thread safe and does not
+/// inherently allow for synchronized access. Callers are responsible for synchronization
+/// (currently using serial dispatch queues).
+@objc(RCNConfigSettings) open class ConfigSettings: NSObject {
+  // MARK: - Private Properties
+
+  /// A list of successful fetch timestamps in seconds.
+  private var _successFetchTimes: [TimeInterval] = []
+
+  /// A list of failed fetch timestamps in seconds.
+  private var _failureFetchTimes: [TimeInterval] = []
+
+  /// Device conditions since last successful fetch from the backend. Device conditions including
+  /// app version, iOS version, device locale, language, GMP project ID and Game project ID.
+  /// Used for determining whether to throttle.
+  private var _deviceContext: [String: Any] = [:]
+
+  /// Custom variables (aka App context digest). This is the pending custom variables
+  /// request before fetching.
+  private var _customVariables: [String: Any] = [:]
+
+  /// Last fetch status.
+  private var _lastFetchStatus: RemoteConfigFetchStatus = .noFetchYet
+
+  /// Last fetch Error.
+  private var _lastFetchError: RemoteConfigError
+
+  /// The time of last apply timestamp.
+  private var _lastApplyTimeInterval: TimeInterval = 0
+
+  /// The time of last setDefaults timestamp.
+  private var _lastSetDefaultsTimeInterval: TimeInterval = 0
+
+  /// The database manager.
+  private var _DBManager: ConfigDBManager
+
+  /// The namespace for this instance.
+  private let _FIRNamespace: String
+
+  /// The Google App ID of the configured FIRApp.
+  private let _googleAppID: String
+
+  /// The user defaults manager scoped to this RC instance of FIRApp and namespace.
+  private var _userDefaultsManager: UserDefaultsManager
+
+  // MARK: - Data required by config request.
+
+  /// InstallationsID.
+  /// - Note: The property is atomic because it is accessed across multiple threads.
+  @objc public var configInstallationsIdentifier: String?
```

Comment:
Does adding `@objc` make it atomic?

Reply:
Good question. It looks like ObjC generated interface uses nonatomic:
<img width="479" alt="Screenshot 2024-12-17 at 10 43 51 AM" src="https://github.com/user-attachments/assets/35fe6b81-67fa-4645-bf5b-f16a2ec2fe7b" />

Reply:
We may need a bigger refactor to get the atomics right across RC. I'm ok to log a TODO for now.